### PR TITLE
add methods to read, format and print rhessi flare list

### PR DESCRIPTION
**Pull request by Laubeee**
Fixes #1018 

considerations:
- using a function (consistent with existing functionality) instead of class (as proposed in the issue)
- requires astropy 4.0 for correct cache handling when downloading multiple files (and also for timedelta.to_value method)

### CODE REVIEW

File: sunpy/instr/rhessi.py [hunk 1]
```diff
@@ -348,3 +400,185 @@ def _build_energy_bands(label, bands):
...
+        The URL or local filename of the hessi flare list.
+
+    Returns
+    -------
+    ``pandas.DataFrame``
+        out : ``pandas.DataFrame`` containing the parsed flares.
+
+    Examples
+    --------
+    >>> from sunpy.instr.rhessi as read_flare_file as rff
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]:
```suggestion
    >>> from sunpy.instr.rhessi import read_flare_file as rff
```
Seems to be be causing a problem.

File: sunpy/instr/rhessi.py [hunk 2]
```diff
@@ -348,3 +400,185 @@ def _build_energy_bands(label, bands):
...
+    >>> rff("https://hesperia.gsfc.nasa.gov/hessidata/dbase/hessi_flare_list_201802.fits")
```
Comment (samaloney) [inline sunpy/instr/rhessi.py]: I think some of the build errors were caused by this, tests which require remote data have to be marked otherwise they fail

```suggestion
    >>> rff("https://hesperia.gsfc.nasa.gov/hessidata/dbase/hessi_flare_list_201802.fits")  # doctest: +REMOTE_DATA
```

File: sunpy/instr/rhessi.py [hunk 3]
```diff
@@ -348,3 +400,185 @@ def _build_energy_bands(label, bands):
...
+    """
+    try:
+        fits = sunpy.io.fits.read(file)
+    except Exception:
+        raise RuntimeError("couldn't load file " + file)
+
+    results = []
+    for row in fits[3].data[0:20]:
+        result_row = {}
+        for k in fits_rec_keys:
```
Comment (samaloney) [inline sunpy/instr/rhessi.py]: I'm probably missing something but could you do `for k in fits[3].data.columns.names` and remove the constant at the top of the file or is than an optimisation?

File: sunpy/instr/rhessi.py [hunk 4]
```diff
@@ -348,3 +400,185 @@ def _build_energy_bands(label, bands):
...
+                x=int(round(row['POSITION'][0])),
+                y=int(round(row['POSITION'][1])),
+                r=int(round((row['POSITION'][0] ** 2 + row['POSITION'][1] ** 2) ** 0.5)),
+                ar=row['ACTIVE_REGION'],
+                f=" ".join(convert_flag_dict(row['FLAGS'])),
+            )
+        )
+
+
+def convert_flag_dict(flags_dict):
```
[the remark this answers is not in the document]
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: Not sure.. I still "made it public" as it could be helpful if you want to do something else with it than print it directly. The raw flags are quite messy.. but yeah, its a rather rare use case i guess

File: sunpy/instr/rhessi.py [hunk 5]
```diff
@@ -348,3 +400,185 @@ def _build_energy_bands(label, bands):
...
+    ----------
+    https://hesperia.gsfc.nasa.gov/rhessi3/data-access/rhessi-data/flare-list/index.html
+    """
+    try:
+        fits = sunpy.io.fits.read(file)
+    except Exception:
+        raise RuntimeError("couldn't load file " + file)
+
+    results = []
+    for row in fits[3].data[0:20]:
```
[the remark this answers is not in the document]
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: oh! good catch. That was due to local testing on my end.

File: sunpy/instr/rhessi.py (hunk 2 above)
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: This is going to be too long (>100char)
Does it work like this too?
```python
>>> # doctest: +REMOTE_DATA
>>> rff("https://hesperia.gsfc.nasa.gov/hessidata/dbase/hessi_flare_list_201802.fits")
```

otherwise I have to do something like this
```python
>>> url = "https://hesperia.gsfc.nasa.gov/hessidata/dbase/hessi_flare_list_201802.fits"
>>> rff(url)  # doctest: +REMOTE_DATA
```

File: sunpy/instr/rhessi.py [hunk 6]
```diff
@@ -27,6 +34,51 @@
 lc_linecolors = ('black', 'pink', 'green', 'blue', 'brown', 'red',
                  'navy', 'orange', 'green')
 
+known_flare_list_sources = {
+    "NASA": "https://hesperia.gsfc.nasa.gov/hessidata/dbase/",
+    "i4DS": "http://soleil.i4ds.ch/hessidata/dbase/",
+}
+
+flagid2flag = {
```
[the remark this answers is not in the document]
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: hm. I named them after the existing `grid_pitch`, `grid_orientation` and `lc_linecolors` ... I suggest to either change all or none to uppercase.. or are they somehow different?

on the other hand, changing them could break behaviour for users.. so if you'd prefer to leave them as-is but still introduce the new ones as all-caps I guess that would be somewhat justified.

File: sunpy/instr/rhessi.py (hunk 2 above)
Comment (samaloney) [reply sunpy/instr/rhessi.py]: I actually don't know the second option should work though

File: sunpy/instr/rhessi.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (samaloney) [reply sunpy/instr/rhessi.py]: I think it would be good to at least change the ones you created. 

A quick search of the code show these are only used from this class and I can't see these also they are not in `__all__` so not exposed also I can see these being  used directly so I would change the rest too, but maybe wait for a second opinion on that.

[the remark this answers is not in the document]
Comment (nabobalis) [reply sunpy/instr/rhessi.py]: Just a quick note, as long as its defined in the code, they are accessible (aka you can import them directly if you look for it). The all just hides them if you do import *.

**Comment by ayshih** [comment]
@Laubeee Please rebase this PR to clean up the merging of unrelated commits.  Let us know if you need help with that.

**Comment by Laubeee** [comment]
> @Laubeee Please rebase this PR to clean up the merging of unrelated commits. Let us know if you need help with that.

Sorry about that. I tried to keep the master updated and didn't know that the commits will then show up in here.. or maybe somewhere i messed up something in the process

If you know how to revert this help would be appreciated, otherwise I will just not update (rebase) this branch anymore

**Comment by ayshih** [comment]
The easiest way in this situation is probably an interactive rebase.  From your local branch, type:
```
git fetch upstream
git rebase -i upstream/master
```
In the editor, you'll see:
```
pick 4903066 add methods to read, format and print rhessi flare list
pick f6b6548 fix pep8 issues about 100char limit
pick c29f671 change date format to sunpy.time
pick 6693d50 add methods to read, format and print rhessi flare list
pick d3b9562 fix pep8 issues about 100char limit
```
The fourth and fifth lines are duplicates of earlier commits due to the merge you performed, so they should be deleted.  You'll be left with:
```
pick 4903066 add methods to read, format and print rhessi flare list
pick f6b6548 fix pep8 issues about 100char limit
pick c29f671 change date format to sunpy.time
```
Save and exit, and your branch is now rebased.  Then, you'll have to force push it back to GitHub:
```
push -f origin rhessi_flarelist
```

**Comment by nabobalis** [comment]
I was going to say, we could do it for you but I don't seem to be able to push.

File: sunpy/instr/rhessi.py (hunk 3 above)
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: Nope, you're absolutely right :) I simply didn't know this exists..

**Comment by Cadair** [comment]
If you haven't already it would be worth comparing this to #1123 (also x-ref for metadata purposes).

**Comment by Laubeee** [comment]
> If you haven't already it would be worth comparing this to #1123 (also x-ref for metadata purposes).

I skimmed through it.. seems the main difference is that one read the full text flarelist instead of monthly and fits.
Also I'm going to add [this detail description](http://sprg.ssl.berkeley.edu/~jimm/hessi/hsi_flare_list.html) to the code / docs comments as it provides helpful information about all the fields.

File: sunpy/instr/rhessi.py [hunk 7]
```diff
@@ -348,3 +391,179 @@ def _build_energy_bands(label, bands):
...
+    Parameters
+    ----------
+    start : `str`
+        Start date of period within which flares should be loaded.
+    end : `str`
+        End date of period within which flares should be loaded.
+    source : `str`, optional
+        Source from where .fits files should be loaded. Can be an URL or a local folder.
+        Known sources are "NASA", "Berkeley" and "i4DS".
+        Defaults to ``"NASA"``.
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]:
```suggestion
        Defaults to "NASA".
```

File: sunpy/instr/rhessi.py [hunk 8]
```diff
@@ -348,3 +391,179 @@ def _build_energy_bands(label, bands):
...
+    file_format : `str`, optional
+        Specifies the naming convention of the files available in the source folder.
+        The changing parts should be given as format string as in ``strftime``.
+        Defaults to ``"hessi_flare_list_%Y%m.fits"``.
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]:
```suggestion
        Defaults to "hessi_flare_list_%Y%m.fits".
```

File: sunpy/instr/rhessi.py [hunk 9]
```diff
@@ -348,3 +391,179 @@ def _build_energy_bands(label, bands):
...
+    inc : `timedelta`, `relativedelta`, optional
+        Specifies by how much time the single files are separated.
+        Defaults to ``relativedelta(months=+1)``.
+
+    Returns
+    -------
+    ``pandas.DataFrame``
+        out : ``pandas.DataFrame`` containing the flares within the given time constraints
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]:
```suggestion
        A dataframe containing the flares within the given time constraints.
```

File: sunpy/instr/rhessi.py [hunk 10]
```diff
@@ -348,3 +391,179 @@ def _build_energy_bands(label, bands):
...
+
+    Parameters
+    ----------
+    file : `str`
+        The URL or local filename of the hessi flare list.
+
+    Returns
+    -------
+    ``pandas.DataFrame``
+        out : ``pandas.DataFrame`` containing the parsed flares.
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]:
```suggestion
        A dataframe containing the parsed flares.
```

File: sunpy/instr/rhessi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nabobalis) [reply sunpy/instr/rhessi.py]: I am not sure if I am keen of these two functions if they have a rare use case. Willing to be persuaded otherwise.

**Comment by Laubeee** [comment]
Seems I broke some (other) tests in the process... anyone an idea why they are affected & how to fix it?

**Comment by nabobalis** [comment]
> Seems I broke some (other) tests in the process... anyone an idea why they are affected & how to fix it?

Since you added some test files, all you need to do to fix these broken tests is to change the answer to the now correct one.

```
    def test_entries_from_dir_recursively_false():
        with pytest.warns(AstropyUserWarning, match='File may have been truncated'):
            entries = list(entries_from_dir(testdir, False,
                                            default_waveunit='angstrom',
                                            time_string_parse_format='%d/%m/%Y'))
>       assert len(entries) == 110
E       assert 122 == 110
E         -122
E         +110

../.tox/py36-offline/lib/python3.6/site-packages/sunpy/database/tests/test_tables.py:419: AssertionError
______________________________________________________________________ test_entries_from_dir_recursively_true ______________________________________________________________________
[gw0] linux -- Python 3.6.9 /root/project/.tox/py36-offline/bin/python

    def test_entries_from_dir_recursively_true():
        with pytest.warns(AstropyUserWarning, match='File may have been truncated'):
            entries = list(entries_from_dir(testdir, True,
                                            default_waveunit='angstrom',
                                            time_string_parse_format='%d/%m/%Y'))
>       assert len(entries) == 131
E       assert 143 == 131
E         -143
E         +131
```

Just update the numbers to match and it should be ok.

File: sunpy/instr/tests/test_rhessi.py [hunk 11]
```diff
@@ -190,3 +191,59 @@ def test_build_energy_bands(raw_bands):
...
+    assert fl.loc[2]['E_PHFLUENCE'] == pytest.approx(12945.400390625, fa)
+    assert fl.loc[3]['PEAK_PHFLUX_SIGMA'] == pytest.approx(37513084.0, fa)
+    assert fl.loc[4]['TOT_PHFLUENCE_SIGMA'] == pytest.approx(354753760.0, fa)
+    assert fl.loc[5]['E_PHFLUENCE_SIGMA'] == pytest.approx(24626434.0, fa)
+
+
+@pytest.mark.remote_data
+def test_get_flare_list_remote():
+    fl = rhessi.get_flare_list("2017-11-16", "2018-02-04")
+    assert len(fl) == pytest.approx(13, 2)  # list may change in the future
```
Comment (nabobalis) [inline sunpy/instr/tests/test_rhessi.py]: This is new to me and I love it.

File: sunpy/instr/rhessi.py [hunk 12]
```diff
@@ -348,3 +391,183 @@ def _build_energy_bands(label, bands):
     unit = matched.group('UNIT').strip()
 
     return [f'{band} {unit}' for band in bands]
+
+
+def get_flare_list(start: str,
+                   end: str,
+                   source: str='NASA',
+                   file_format: str="hessi_flare_list_%Y%m.fits",
+                   inc=relativedelta(months=+1)):
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]: Stupid question, why the +1?

File: sunpy/instr/rhessi.py [hunk 13]
```diff
@@ -348,3 +391,183 @@ def _build_energy_bands(label, bands):
...
+    """
+    Read and combine RHESSI flare lists from .fits files as specified with further parameters
+    Supported date formats are the same as in ``sunpy.time``.
```
Comment (nabobalis) [inline sunpy/instr/rhessi.py]: We have a way to add time types to the docs. 

See https://github.com/sunpy/sunpy/blob/1f79409372f8bd932c8f4db978e67bff451f04a7/sunpy/coordinates/sun.py#L45 for an example.

File: sunpy/instr/rhessi.py (hunk 12 above)
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: Well, I want to iterate through every month... the + is optional, but it seems to be preferred syntax. I guess it's clearer whether you want to increase or decrease a date this way..

File: sunpy/instr/rhessi.py (hunk 13 above)
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: I hope I used that somewhat correctly, but it seems it triggered an error in the tests (`timeseries_example.py`):

```
Warning, treated as error:
/home/circleci/project/examples/time_series/timeseries_example.py failed to execute correctly: Traceback (most recent call last):
  File "/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/home/circleci/project/examples/time_series/timeseries_example.py", line 144, in <module>
    df_downsampled = ts_goes_trunc.data.resample('10T', 'mean')
  File "/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/pandas/core/generic.py", line 8102, in resample
    axis = self._get_axis_number(axis)
  File "/home/circleci/project/.tox/build_docs/lib/python3.7/site-packages/pandas/core/generic.py", line 407, in _get_axis_number
    raise ValueError(f"No axis named {axis} for object type {cls}")
ValueError: No axis named mean for object type <class 'pandas.core.frame.DataFrame'>
```

**Comment by Laubeee** [comment]
> I was going to say, we could do it for you (if you prefer) but I don't seem to be able to push.

Didn't realize where this setting was. Changed it ;)

File: sunpy/instr/rhessi.py (hunk 13 above)
Comment (nabobalis) [reply sunpy/instr/rhessi.py]: That seems to be the new version of pandas that we had to update the example for. If you can merge in master, hopefully that should go away.

Comment (Laubeee) [reply sunpy/instr/rhessi.py]: Oh ok. Should I try to rebase my fork?

Comment (nabobalis) [reply sunpy/instr/rhessi.py]: You can either rebase the fork if you are comfortable doing that, or you can update the remote that points to this repository (if you have one) and then do `git pull <name> master`.

Comment (Laubeee) [reply sunpy/instr/rhessi.py]: phew. That resolved the error but introduced new ones :'(
also locally i cant do `import sunpy` anymore, it complains
```
File "C:\Projects\JHV\Python\sunpy\sunpy\__init__.py", line 18, in <module>
    from .version import __version__
ModuleNotFoundError: No module named 'sunpy.version'
```

it seems someone added `sunpy/version.py` to the gitignore..

Comment (nabobalis) [reply sunpy/instr/rhessi.py]: Yeah one of the changes in master, was to generate the version.py file at install. So you now how have to `pip install -e .` to create it. I am not sure we documented it anywhere (but it was how it used to be about 2 years ago).

**Comment by Laubeee** [comment]
Further considerations:
- `print_flare_list` will round the position numbers, as opposed to just cutting the decimals of (as it seems to be the case in the `.txt` lists)
- The Active Region in the `.fits` files is sometimes different from the one in the `.txt` files. Sometimes only one of them lists an AR, sometimes both list a different one. I observed this especially for [1st-3rd November 2003](https://hesperia.gsfc.nasa.gov/hessidata/dbase/hessi_flare_list_200311.txt). If anyone has an idea why that is, we could add a note in the docs about it (or make the files match, for that matter.)

**Comment by Laubeee** [comment]
... So, can this be merged? anything missing?

**Comment by nabobalis** [comment]
> ... So, can this be merged? anything missing?

It will need reviews from more people before anything else can happen. Hopefully the other maintainers can get around to this soon. Sorry for the wait!

**Comment by Laubeee** [comment]
No worries, just unsure what the status is here, and whether its me everyone is waiting for :D

**Comment by hayesla** [comment]
One general comment - and up for discussion - is that it may be better to return the flare list as an `astropy.table` rather than a `pandas.DataFrame` - it might just fit in line better with the HEK query search for things like searching for the GOES flare list

```
>>> from sunpy.net import hek
>>> client = hek.HEKClient()
>>> event_type = 'FL' 
>>> result=client.search(hek.attrs.Time('2013-10-28', '2013-10-29'),hek.attrs.EventType(event_type),hek.attrs.OBS.Observatory == 'GOES') 
>>> result
<HEKTable length=9>
         SOL_standard          active ... skel_startc2 sum_overlap_scores
            str30               str4  ...    object           str1       
------------------------------ ------ ... ------------ ------------------
SOL2013-10-28T01:25:00L024C086   true ...         None                  0
SOL2013-10-28T01:41:00L031C086   true ...         None                  0
.....

```
Where the base class of the HEK table is an astropy.table.Table

Thoughts?

File: sunpy/instr/rhessi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: Do I understand correct, you would return both the raw and the formatted flags in the DF?
I think there are 1 or 2 flags that are not present in the formatted version, which is why I think access to the raw flags should not be lost. However they also don't seem to be of much interest to the people that use this data, otherwise it would be included in the .txt version of the flare lists...

File: sunpy/instr/rhessi.py [hunk 14]
```diff
@@ -348,3 +395,195 @@ def _build_energy_bands(label, bands):
...
+                    flags[fid] = i
+                result_row[k] = flags
+            else:
+                result_row[k] = row[k]
+        results.append(result_row)
+
+    return pd.DataFrame(results)
+
+
+def print_flare_list(data_frame):
```
[the remark this answers is not in the document]
Comment (Laubeee) [reply sunpy/instr/rhessi.py]: this function outputs the flare in (almost) the same way they are presented in the .txt versions of the flare lists. I used it to compare my data extracted from .fits to the .txt lists (to see if it matches or if I have errors..) and thought someone else might want to use it too..

File: sunpy/instr/rhessi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wtbarnes) [reply sunpy/instr/rhessi.py]: Sorry, should've clarified. Yes, I meant return both the raw and formatted flags so that the raw data is not lost and so that the user does not have to call an additional function to get the formatted ones. I have not worked with this data at all so others with more experience may have different opinions.

**Comment by Laubeee** [comment]
Hey @hayesla , hey @wtbarnes 
Thanks for your comments
I am willing to refactor this from pandas DF to astropy and put it in a separate class, if this you, in the sunpy maintainer group, are supporting this on a broad level.

As for the usefulness of the functionality: it is my plan to introduce this (and a potential future PR that can convert RHESSI image cubes to a series of 2D fits maps (IDL hsi_fits2map)) to my colleagues in the heliophysics team. The use case is that they can, with as few effort as possible, import the images corresponding to a specific flare into JHV for visualization (and overlaying with other sources). It is currently not urgent but for them it would obviously be nice to have it in an official sunpy release, so as long as it can go into 2.0 (which is set for May at this time) that'd be great for me.

File: sunpy/instr/rhessi.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (wtbarnes) [reply sunpy/instr/rhessi.py]: Fair enough though I'm not really in favor of including functions that just print information to the screen. I'd suggest removing this function and adding any additional derived quantities (e.g. it looks like you're calculating the duration, `dur`, based on the start and end times) to the resulting dataframe. That way, these useful quantities are still exposed to the user, but the data representation is offloaded to the dataframe (or astropy Table, if we decide to go that direction).

**Comment by Laubeee** [comment]
So, I finally managed to get back to this.
I changed all the small changes. For the HekClient-like interface, there is a high chance I can do this in the coming weeks

Just to wrap my head around what it actually means more in detail:
- wrap it all in a class `RHESSIFlareClient`
- rename `get_flare_list` to `search` and pass params as `attrs` (a bit unsure what it actually means in terms of coding.. but I will probably figure it out)
- return data as `astropy.table` (or RHESSIFlareTable, if needed. Maybe we could put the formatting functionality in there?)

- Question: Do I understand correct: you want to move the functionality into sunpy.net? Should I create a new file `rhessi.py` there?

**Comment by Laubeee** [comment]
update from my side on this PR:
While it should already be functional in reading and extracting rhessi flare list fits files, my work on it is kind of discontinued, since there was a decision to make the flare list also available through VESPA-DaCHS for maximum interoperability and therefore can be loaded through [pyvo](https://github.com/astropy/pyvo). (side note: this works already on internal systems, but it will be a couple of months until the service is ready to be publicly available and registred to VESPA, also due to the ongoing work regarding the rhessi flare image archive being still under construction at GSFC)

So in short: It is unlikely I will be able to restructure this PR according to your inputs in the near future. Of course anyone willing could pick it up or it could even be integrated as-is if the functionality is still considered useful.

**Comment by nabobalis** [comment]
Is there any chance you can open this PR on the sunkit-instruments repository?

**Comment by Laubeee** [comment]
If desired...
As I said, the flare list can in the future be browsed through VESPA / pyvo, so I'm not quite sure there is still a use case for this?

**Comment by hayesla** [comment]
actually following up on this - there is actually going to be an example in the gallery of how to query the rhessi flare list from helio so may not need this

see https://github.com/sunpy/sunpy/blob/850b18924fc4059e0082e5d79d8ec6216c61c075/examples/acquiring_data/fido_metadata_queries.py in #4358

**Comment by Laubeee** [comment]
Ok, I'm closing this here. I'll post a pyvo example in the issue as soon as that method is publicly available, so the issue can be closed as well. If anyone is still interested in this code I'll re-open a PR in the SKI-repo.
Maybe @ehsteve who initially opened the issue has opinions?

**Comment by kakirastern** [comment]
Thanks @hayesla for following up for us!
